pkg/artifactory/provider.go

```diff
@@ -164,10 +174,59 @@ func providerConfigure(d *schema.ResourceData) (interface{}, error) {
 	commandid := "Terraform/" + version.Version
 	usage.SendReportUsage(productid, commandid, rtnew)
 
+	var xrayClient *xray.XrayServicesManager
+	if d.Get("xray_url") != nil && d.Get("xray_url").(string) != "" {
+		xrayClient, err = createXrayClient(d)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	rt := &ArtClient{
-		ArtOld: rtold,
-		ArtNew: rtnew,
+		ArtOld:     rtold,
+		ArtNew:     &rtnew,
+		XrayClient: xrayClient,
 	}
 
 	return rt, nil
 }
+
+func createXrayClient(d *schema.ResourceData) (*xray.XrayServicesManager, error) {
+	details := xrayauth.NewXrayDetails()
+
+	username := d.Get("username").(string)
+	password := d.Get("password").(string)
+	apiKey := d.Get("api_key").(string)
+	accessToken := d.Get("access_token").(string)
+
+	url := d.Get("xray_url").(string)
+	if url[len(url)-1] != '/' {
+		url += "/"
+	}
+	details.SetUrl(url)
+
+	if username != "" && password != "" {
```

Comment:
These are not nested cases (so, no `else if` needed).

Also, please have independent cases for element missing; not all bunched up into 1 error

Reply:
Hi @chb0github ,

There are 3 ways someone could authenticate to xray.
1. User name and password
1. API Key
1. Access token

The cases are mutually exclusive.
Hence, the `else if`.